### PR TITLE
Enable predicate invention from txt files

### PR DESCRIPTION
**Pull request by NishanthJKumar**
This PR enables our typical grammar search invention to take in carefully formatted `txt` files as input. This builds towards enabling the pipeline to operate directly over data supplied by a VLM.

### CODE REVIEW

File: predicators/datasets/vlm_input_data.py [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+
+
+def create_ground_atom_data_from_labeled_txt(
+        env: BaseEnv, train_tasks: List[Task],
+        known_options: Set[ParameterizedOption]) -> Dataset:
+    """Given a txt file containing trajectories labelled with VLM predicate
+    values, construct a dataset that can be passed to the rest of our learning
+    pipeline."""
+    dataset_fpath = os.path.join(CFG.data_dir, CFG.handmade_demo_filename)
+    # First, parse this dataset into a structured form.
```
[the remark this answers is not in the document]
Comment (tomsilver) [reply predicators/datasets/vlm_input_data.py:225]: after reading the parsing, I'm undecided, will leave it up to you

File: predicators/datasets/vlm_input_data.py [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+
+
+def _parse_structured_actions_into_ground_options(
+        structuredactions_trajs: List[List[Tuple[str, Tuple[str, ...],
+                                                 List[float]]]],
+        known_options: Set[ParameterizedOption],
+        train_tasks: List[Task]) -> List[List[_Option]]:
+    """Convert structured actions trajectories into actual lists of ground
+    options trajectories."""
+    assert len(structuredactions_trajs) == len(train_tasks)
```
Comment (tomsilver) [inline predicators/datasets/vlm_input_data.py:125]: grep for structuredactions -> structured_actions

File: predicators/datasets/vlm_input_data.py [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+            curr_option_traj.append(ground_option)
+        option_trajs.append(curr_option_traj)
+    return option_trajs
+
+
+def _create_dummy_goal_state_for_each_task(
+        env: BaseEnv, train_tasks: List[Task]) -> List[State]:
+    """Uses a lot of assumptions to generate a state in which a dummy goal
+    predicate holds for each train task."""
+    # We assume there is only one goal predicate, and that it is
```
Comment (tomsilver) [inline predicators/datasets/vlm_input_data.py:153]: is this going to be a long-term assumption?

File: predicators/datasets/vlm_input_data.py [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+    # NOTE: In this LowLevelTrajectory, we assume the low level states
+    # are the same as the init state until the final state.
+    trajs = []
+    for traj_num in range(len(option_trajs)):
+        traj_init_state = train_tasks[traj_num].init
+        curr_traj_states = []
+        curr_trajactions = []
+        for idx_within_traj in range(len(option_trajs[traj_num])):
+            curr_traj_states.append(traj_init_state)
+            curr_trajactions.append(
```
Comment (tomsilver) [inline predicators/datasets/vlm_input_data.py:193]: `traj_actions`

File: predicators/envs/vlm_envs.py [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+
+    def _generate_train_tasks(self) -> List[EnvironmentTask]:
+        return self._get_tasks(num=CFG.num_train_tasks, rng=self._train_rng)
+
+    def _generate_test_tasks(self) -> List[EnvironmentTask]:
+        return self._get_tasks(num=CFG.num_test_tasks, rng=self._test_rng)
+
+    @property
+    def predicates(self) -> Set[Predicate]:
+        return set([self._DummyGoal])
```
Comment (tomsilver) [inline predicators/envs/vlm_envs.py:59]: nit: `{self._DummyGoal}`

File: predicators/envs/vlm_envs.py [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+        self._object_type = Type("object", [])
+        self._goal_object_type = Type("goal_object", ["goal_true"],
+                                      self._object_type)
+
+        # Predicates
+        self._DummyGoal = Predicate("DummyGoal", [self._goal_object_type],
+                                    self._Dummy_Goal_holds)
+
+    def simulate(self, state: State, action: Action) -> State:
+        raise ValueError("simulate shouldn't be getting called!")
```
Comment (tomsilver) [inline predicators/envs/vlm_envs.py:40]: nit: for consistency, capitalize exception messages

File: predicators/envs/vlm_envs.py [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+        """Return a set of atom strings that should be sufficient for a VLM to
+        label demonstrations consistently to learn good operators."""
+        raise NotImplementedError(
+            "VLM debug atom strings not implemented for this environment.")
+
+    def _get_tasks(
+        self, num: int, rng: np.random.Generator
+    ) -> List[EnvironmentTask]:  # pragma: no cover.
+        del num, rng
+        raise NotImplementedError("Override!")
```
Comment (tomsilver) [inline predicators/envs/vlm_envs.py:89]: can this method just be removed, if this environment is meant to still be ABC?

File: predicators/envs/vlm_envs.py [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+
+
+class IceTeaMakingEnv(VLMPredicateEnv):
+    """An environment to test demonstrations involving making a cup of tea."""
```
Comment (tomsilver) [inline predicators/envs/vlm_envs.py:93]: "to test demonstrations" sounds weird to me

File: predicators/envs/vlm_envs.py [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+        # Env-specific types.
+        self._teabag_type = Type("teabag", [], self._object_type)
+        self._spoon_type = Type("spoon", [], self._object_type)
+        self._cup_type = Type("cup", [], self._object_type)
+        self._plate_type = Type("plate", [], self._object_type)
+        self._hand_type = Type("hand", [], self._object_type)
+
+    @classmethod
+    def get_name(cls) -> str:
+        return "iced_tea_making"
```
Comment (tomsilver) [inline predicators/envs/vlm_envs.py:107]: iced tea or ice tea?

File: predicators/envs/vlm_envs.py [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+            state: State,
+            task: EnvironmentTask,
+            action: Optional[Action] = None,
+            caption: Optional[str] = None) -> matplotlib.figure.Figure:
+        raise ValueError("shouldn't be trying to render env at any point!")
+
+    @property
+    @abc.abstractmethod
+    def get_vlm_debug_atom_strs(self) -> Set[str]:
+        """Return a set of atom strings that should be sufficient for a VLM to
```
Comment (tomsilver) [inline predicators/envs/vlm_envs.py:80]: can you say a little more about this? why is this needed?

**Review by tomsilver — COMMENTED** [review]
looks great! awesome work. all minor comments. one question: how are you dealing with object types in parsing? at first I thought you were using all generic `object` typing, which would make sense, but then I saw your ice tea environment has specific types in it?

**Comment by NishanthJKumar** [comment]
Thanks!

We assign predicates to have types corresponding to the names of objects from the environment (e.g. if the txt file has `OnTop(teabag, cup)`, then we will make a predicate `OnTop(?teabag:teabag_type, ?cup:cup_type)`). However, if there are multiple `OnTop` predicates with different objects in this case, we'll just fallback to the generic `object` type. I like this behavior because we can always make it default to objects by not declaring any types in the environment, but we also get finer-grained control as and when we want

File: predicators/approaches/documentation/grammar_search_invention_approach.md [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+place_in(spoon, cup)[] -> 
+{*Holding(spoon): True.
+*Submerged(teabag): False.
+*Submerged(spoon): False.}
+===
+```
+
+**Components**
+- Separator: '===' is used to separate one trajectory from another (so a trajectory is sandwiched between two lines that have only '===' on them). In the above example, there is exactly one demonstration trajectory.
+- State: Each state is a bulleted list of atoms enclosed between set brackets {}. In the above example, there are 5 states. Note importantly that the format of every atom should be `*<predicate_name>(<ob1_name>, <obj2_name>, ...).`. The `*` at the start, and the period `.` at the end are very important.
```
[the remark this answers is not in the document]
Comment (NishanthJKumar) [reply predicators/approaches/documentation/grammar_search_invention_approach.md:43]: actually `*` isn't necessary. `.` resolves ambiguity for the VLM in cases where it decides to output explanations for why it sets predicates to certain values (i.e., chain of thought)

File: predicators/envs/vlm_envs.py (hunk 7 above)
Comment (NishanthJKumar) [reply predicators/envs/vlm_envs.py:89]: I did that originally, but the linter started complaining :/

File: predicators/datasets/vlm_input_data.py (hunk 3 above)
Comment (NishanthJKumar) [reply predicators/datasets/vlm_input_data.py:153]: no - just for now!

File: predicators/envs/vlm_envs.py (hunk 10 above)
Comment (NishanthJKumar) [reply predicators/envs/vlm_envs.py:80]: Oops - this is more of a VLM thing; I'll remove from this PR

File: predicators/datasets/vlm_input_data.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (NishanthJKumar) [reply predicators/datasets/vlm_input_data.py:225]: So the real reason for this whole structured thing is that it allows good code sharing between all the VLM parsing code that's about to be in the next PR, and the code that parses these txt files. Basically, the flow looks like this:
**txt files**
input txt -> structured dicts -> regular data structures
**vlm**
input image trajs -> structured dicts -> regular data structures
So we get to share all the code that does `structured dicts -> regular data structures`. What's more, the mapping between input txt and structured dicts is reversible (i.e., we can easily recover the input txt from the structured txt). This will let us dump out txt's from the VLM part of the pipeline, which is super helpful!

File: predicators/approaches/grammar_search_invention_approach.py [hunk 12]
```diff
@@ -924,44 +924,66 @@ def _get_current_predicates(self) -> Set[Predicate]:
...
-            f"_{CFG.grammar_search_grammar_use_diff_features}" + \
-            f"_{CFG.grammar_search_use_handcoded_debug_grammar}" + \
-            dataset_fname[-5:]
-
-        # Load pre-saved data if the CFG.load_atoms flag is set.
-        atom_dataset: Optional[List[GroundAtomTrajectory]] = None
-        if CFG.load_atoms:
-            atom_dataset = utils.load_ground_atom_dataset(
-                dataset_fname, dataset.trajectories)
+        if not CFG.offline_data_method == "demo+labeled_atoms":
```
[the remark this answers is not in the document]
Comment (NishanthJKumar) [reply predicators/approaches/grammar_search_invention_approach.py:927]: Not exactly sure what you mean here? Followed up on Slack

**Review by tomsilver — APPROVED** [review]
